test/integration/src/test/java/io/pravega/test/integration/RestoreBackUpDataRecoveryTest.java

```diff
@@ -0,0 +1,648 @@
+                if (dataLogFactory != null) {
+                    this.serviceBuilder = ServiceBuilder.newInMemoryBuilder(ServiceBuilderConfig.getDefaultConfig())
+                            .withStorageFactory(setup -> storageFactory)
+                            .withDataLogFactory(setup -> dataLogFactory);
+                } else {
+                    this.serviceBuilder = ServiceBuilder.newInMemoryBuilder(ServiceBuilderConfig.getDefaultConfig())
+                            .withStorageFactory(setup -> storageFactory);
+                }
+            } else {
+                this.serviceBuilder = ServiceBuilder.newInMemoryBuilder(ServiceBuilderConfig.getDefaultConfig());
+            }
+            this.serviceBuilder.initialize();
+            this.streamSegmentStoreWrapper = new StreamSegmentStoreWrapper(serviceBuilder.createStreamSegmentService());
+            this.monitor = new AutoScaleMonitor(streamSegmentStoreWrapper, AutoScalerConfig.builder().build());
+            this.tableStoreWrapper = new TableStoreWrapper(serviceBuilder.createTableStoreService());
+            this.server = new PravegaConnectionListener(false, false, "localhost", servicePort, streamSegmentStoreWrapper,
+                    tableStoreWrapper, monitor.getStatsRecorder(), monitor.getTableSegmentStatsRecorder(), new PassingTokenVerifier(),
+                    null, null, true, serviceBuilder.getLowPriorityExecutor());
+            this.server.startListening();
+        }
+
+        public void close() {
+            if (this.server != null) {
+                this.server.close();
+                this.server = null;
+            }
+
+            if (this.monitor != null) {
+                this.monitor.close();
+                this.monitor = null;
+            }
+
+            if (this.serviceBuilder != null) {
+                this.serviceBuilder.close();
+                this.serviceBuilder = null;
+            }
+        }
+    }
+
+    ControllerStarter startController(int bkPort, int servicePort) throws InterruptedException {
+        return new ControllerStarter(bkPort, servicePort);
+    }
+
+    /**
+     * Creates a controller instance and runs it.
+     */
+    private static class ControllerStarter {
+        private final int controllerPort = TestUtils.getAvailableListenPort();
+        private final String serviceHost = "localhost";
+        private ControllerWrapper controllerWrapper = null;
+        private Controller controller = null;
+
+        ControllerStarter(int bkPort, int servicePort) throws InterruptedException {
+            this.controllerWrapper = new ControllerWrapper("localhost:" + bkPort, false,
+                    controllerPort, serviceHost, servicePort, CONTAINER_COUNT);
+            this.controllerWrapper.awaitRunning();
+            this.controller = controllerWrapper.getController();
+        }
+
+        public void close() throws Exception {
+            if (this.controller != null) {
```

Comment:
Verify that you are shutting down everything that needs to shut down.

Reply:
Done.